### PR TITLE
CDI 568

### PR DESCRIPTION
**Pull request by antoinesd**
Unofficial PR to discuss the API

### CODE REVIEW

File: api/src/main/java/javax/enterprise/inject/spi/builder/CDIContainerBuilder.java [hunk 1]
```diff
@@ -0,0 +1,369 @@
...
+     * @return self
+     */
+    CDIContainerBuilder setClassLoader(ClassLoader classLoader);
+
+    /**
+     * <p>
+     * Initializes a CDI Container.
+     * </p>
+     * <p>
+     * Cannot be called within an application server.
```
Comment (mkouba) [inline api/src/main/java/javax/enterprise/inject/spi/builder/CDIContainerBuilder.java]: Not this method but the whole `CDIContainerBuilder` class should not be used in EE, right?

File: api/src/main/java/javax/enterprise/inject/spi/builder/CDIContainerBuilder.java [hunk 2]
```diff
@@ -0,0 +1,369 @@
...
+     */
+    CDI<Object> initialize(Map<String, Object> params);
+
+    /**
+     * Determines whether or not this CDIProvider has been initialized or not.
+     *
+     * @return true if initialized, false if not.
+     * @since 2.0
+     */
+    boolean isInitialized();
```
Comment (mkouba) [inline api/src/main/java/javax/enterprise/inject/spi/builder/CDIContainerBuilder.java]: What's the purpose of this method?

File: api/src/main/java/javax/enterprise/inject/spi/builder/CDIContainerBuilder.java [hunk 3]
```diff
@@ -0,0 +1,369 @@
...
+     * <p>
+     * Cannot be called within an application server.
+     * </p>
+     *
+     * @param params optional parameters, may be null or empty.  May also be immutable.
+     * @return the {@link CDI} instance associated with the container.  This is the same instance returned by using
+     * {@link CDI#current()}
+     * @throws UnsupportedOperationException if called within an application server
+     */
+    CDI<Object> initialize(Map<String, Object> params);
```
Comment (mkouba) [inline api/src/main/java/javax/enterprise/inject/spi/builder/CDIContainerBuilder.java]: We have `property()`/`properties()` methods so I think the param is not needed.

File: api/src/main/java/javax/enterprise/inject/spi/builder/CDIContainerBuilder.java [hunk 4]
```diff
@@ -0,0 +1,369 @@
...
+ */
+public interface CDIContainerBuilder {
+
+    /**
+     * Containers must have a unique identifier assigned when running multiple CDI instances at the same time.
+     *
+     * @param containerId The container identifier
+     * @return self
+     */
+    CDIContainerBuilder containerId(String containerId);
```
Comment (mkouba) [inline api/src/main/java/javax/enterprise/inject/spi/builder/CDIContainerBuilder.java]: I thought you've removed the support for multiple containers?

File: api/src/main/java/javax/enterprise/inject/spi/builder/CDIContainerBuilder.java [hunk 5]
```diff
@@ -0,0 +1,369 @@
...
+
+/**
+ * CDI container builder for Java SE
+ * <p>
+ * <p>
+ * Typical usage looks like this:
+ * </p>
+ * <p>
+ * <pre>
+ * CDI<Object> cdi = Builders.CDIContainer().initialize();
```
Comment (mkouba) [inline api/src/main/java/javax/enterprise/inject/spi/builder/CDIContainerBuilder.java]: We should follow the naming conventions here - `Builders.container()` might be consistent with other public static methods on `Builders`?

File: api/src/main/java/javax/enterprise/inject/spi/builder/CDIContainerBuilder.java [hunk 6]
```diff
@@ -0,0 +1,369 @@
...
+ * CDIContainerBuilder builder = Builders.CDIContainer()
+ *    .disableDiscovery()
+ *    .packages(Main.class, Utils.class)
+ *    .interceptors(TransactionalInterceptor.class)
+ *    .property("property", true);
+ * CDI<Object> cdi = builder.initialize();
+ * </pre>
+ * <p>
+ * <p>
+ * This builder is not reusable.
```
Comment (mkouba) [inline api/src/main/java/javax/enterprise/inject/spi/builder/CDIContainerBuilder.java]: I see no reason why it should not be reusable. Maybe we should remove this sentence.

File: api/src/main/java/javax/enterprise/inject/spi/builder/CDIContainerBuilder.java (hunk 2 above)
Comment (antoinesd) [inline api/src/main/java/javax/enterprise/inject/spi/builder/CDIContainerBuilder.java]: to know if the builder has been already used to build a container.

Comment (mkouba) [inline api/src/main/java/javax/enterprise/inject/spi/builder/CDIContainerBuilder.java]: And what's this info for? Also it shouldn't be thread-safe.

File: api/src/main/java/javax/enterprise/inject/spi/builder/CDIContainerBuilder.java (hunk 6 above)
Comment (antoinesd) [inline api/src/main/java/javax/enterprise/inject/spi/builder/CDIContainerBuilder.java]: Yes. In fact the builder could be reusable. It's a an open point I wanted to discuss with you. I knwo that Mark is against this idea.

File: api/src/main/java/javax/enterprise/inject/spi/builder/CDIContainerBuilder.java (hunk 5 above)
Comment (antoinesd) [inline api/src/main/java/javax/enterprise/inject/spi/builder/CDIContainerBuilder.java]: +1. I also have concern about the builder class: `CDIContainerBuilder` is perhaps a bit too long

**Pull request closed without merging**